Makefile.include

```diff
@@ -4,23 +4,43 @@
 all:
 
 # set undefined variables
-RIOTBASE ?= $(shell dirname "$(lastword $(MAKEFILE_LIST))")
-RIOTBASE := $(abspath $(RIOTBASE))
-
-CCACHE_BASEDIR := $(RIOTBASE)
-
-RIOTCPU ?= $(RIOTBASE)/cpu
-RIOTCPU := $(abspath $(RIOTCPU))
-
-RIOTBOARD ?= $(RIOTBASE)/boards
-RIOTBOARD := $(abspath $(RIOTBOARD))
-
-RIOTPKG ?= $(RIOTBASE)/pkg
+RIOTBASE       ?= $(dir $(lastword $(MAKEFILE_LIST)))
+CCACHE_BASEDIR ?= $(RIOTBASE)
+RIOTCPU        ?= $(RIOTBASE)/cpu
+RIOTBOARD      ?= $(RIOTBASE)/boards
+RIOTPKG        ?= $(RIOTBASE)/pkg
+RIOTPROJECT    ?= $(shell git rev-parse --show-toplevel 2>/dev/null || pwd)
+GITCACHE       ?= $(RIOTBASE)/dist/tools/git/git-cache
+APPDIR         ?= $(CURDIR)
+BINDIRBASE     ?= $(APPDIR)/bin
+BINDIR         ?= $(BINDIRBASE)/$(BOARD)
+
+__DIRECTORY_VARIABLES := RIOTBASE CCACHE_BASEDIR RIOTCPU RIOTBOARD RIOTPKG GITCACHE RIOTPROJECT APPDIR BINDIRBASE BINDIR
+
+# Ensure that all directories are set and don't contain spaces.
+ifneq (, $(filter-out 1, $(foreach v,${__DIRECTORY_VARIABLES},$(words ${${v}}))))
+  $(info Aborting compilation for your safety.)
+  $(info Related variables = ${__DIRECTORY_VARIABLES})
+  $(error Make sure no path override is empty or contains spaces!)
```

Comment:
If there are spaces in the path, yes, that's the point. :) Otherwise Windows' directory layout got smarter, e.g. $HOME is C:\Users\$ID, Eigene Daten is actually $HOME/Documents. There should be no spaces in your common folders unless you put one in there.